ckanext/switzerland/helpers/validators.py

```diff
@@ -442,29 +439,37 @@ def ogdch_validate_formfield_temporals(field, schema):
     This validator is only used for form validation.
     The data is extracted form the temporals form fields and transformed
     into a form that is expected for database storage:
-    "temporals": [{"start_date": 0123456789, "end_date": 0123456789}]
+    "temporals": [{"start_date": <date as isodate>,
+    "end_date": <date as isodate>}]
     """
-    def validator(key, data, errors, context):
-        extras = data.get(FORM_EXTRAS)
-        temporals = []
-        if extras:
-            temporals = get_temporals_from_form(extras)
-            for temporal in temporals:
-                if not temporal['start_date'] and temporal['end_date']:
-                    raise df.Invalid(
-                        _('A valid temporal must have both start and end date')  # noqa
-                    )
-                for value in temporal['start_date'], temporal['end_date']:
-                    if DATE_FORMAT_PATTERN.match(value) is None:
-                        errors[key].append(
-                            _('Expecting DD.MM.YYYY, got "%s"') % str(value)
-                        )
-                temporal['start_date'] = date_string_to_timestamp(temporal['start_date'])  # noqa
-                temporal['end_date'] = date_string_to_timestamp(temporal['end_date'])  # noqa
-        if temporals:
-            data[key] = json.dumps(temporals)
-        elif not _jsondata_for_key_is_set(data, key):
+    def validator(key, data):
+        if key not in data:
```

Comment:
in general here is a bunch of if-else-clauses containing more if-else-clauses. maybe you can consider writing this in a more pythonic way to make it easier to understand?
e.g. what I roughly mean is:
try:
   setting temporals like everything that we expect is here
except:
   set temporals to data[key]

keyword: ask for forgiveness instead of permission https://stackoverflow.com/a/12270384/3507636

Reply:
I spent hours on trying to refactor is. I am normally a big fan of refactoring. But not in this case. Some of the validators are quite complicated and hard to understand. I will refactor it as soon as we have tests for this. Currently we don't have them. Therefore after hours of debugging my refactored version, I decided to leave it as it is.

Reply:
This morning I found out what is really broken there: it goes through two validators, see: https://github.com/opendata-swiss/ckanext-switzerland-ng/blob/6871f45380652ef4a8e6fd00672227d5bce99dd3/ckanext/switzerland/presets.json#L20
The first one `harvest_list_of_dicts` errors, but this is not related to the data change: it is the same error for all our complex field validators: 

See here for contact points for example:
```
ogdch_app     | [ckanext.switzerland.helpers.validators] validator [157] (u'contact_points',)
ogdch_app     | Traceback (most recent call last):
ogdch_app     |   File "/usr/lib/python2.7/logging/__init__.py", line 868, in emit
ogdch_app     |     msg = self.format(record)
ogdch_app     |   File "/usr/lib/python2.7/logging/__init__.py", line 741, in format
ogdch_app     |     return fmt.format(record)
ogdch_app     |   File "/usr/lib/python2.7/logging/__init__.py", line 465, in format
ogdch_app     |     record.message = record.getMessage()
ogdch_app     |   File "/usr/lib/python2.7/logging/__init__.py", line 325, in getMessage
ogdch_app     |     msg = str(self.msg)
ogdch_app     |   File "/usr/lib/ckan/venv/src/ckan/ckan/lib/navl/dictization_functions.py", line 19, in __str__
ogdch_app     |     raise Invalid(_('Missing value'))
ogdch_app     | Invalid: Invalid: u'Missing value'
ogdch_app     | Logged from file validators.py, line 158
ogdch_app     | Traceback (most recent call last):
ogdch_app     |   File "/usr/lib/python2.7/logging/handlers.py", line 76, in emit
ogdch_app     |     if self.shouldRollover(record):
ogdch_app     |   File "/usr/lib/python2.7/logging/handlers.py", line 156, in shouldRollover
ogdch_app     |     msg = "%s\n" % self.format(record)
ogdch_app     |   File "/usr/lib/python2.7/logging/__init__.py", line 741, in format
ogdch_app     |     return fmt.format(record)
ogdch_app     |   File "/usr/lib/python2.7/logging/__init__.py", line 465, in format
ogdch_app     |     record.message = record.getMessage()
ogdch_app     |   File "/usr/lib/python2.7/logging/__init__.py", line 325, in getMessage
ogdch_app     |     msg = str(self.msg)
ogdch_app     |   File "/usr/lib/ckan/venv/src/ckan/ckan/lib/navl/dictization_functions.py", line 19, in __str__
ogdch_app     |     raise Invalid(_('Missing value'))
ogdch_app     | Invalid: Invalid: u'Missing value'
ogdch_app     | Logged from file validators.py, line 158
```

Exactly the same happens for the temporals:
```
ogdch_app     | [ckanext.switzerland.helpers.validators] validator [157] (u'temporals',)
ogdch_app     | Traceback (most recent call last):
ogdch_app     |   File "/usr/lib/python2.7/logging/__init__.py", line 868, in emit
ogdch_app     |     msg = self.format(record)
ogdch_app     |   File "/usr/lib/python2.7/logging/__init__.py", line 741, in format
ogdch_app     |     return fmt.format(record)
ogdch_app     |   File "/usr/lib/python2.7/logging/__init__.py", line 465, in format
ogdch_app     |     record.message = record.getMessage()
ogdch_app     |   File "/usr/lib/python2.7/logging/__init__.py", line 325, in getMessage
ogdch_app     |     msg = str(self.msg)
ogdch_app     |   File "/usr/lib/ckan/venv/src/ckan/ckan/lib/navl/dictization_functions.py", line 19, in __str__
ogdch_app     |     raise Invalid(_('Missing value'))
ogdch_app     | Invalid: Invalid: u'Missing value'
ogdch_app     | Logged from file validators.py, line 158
ogdch_app     | Traceback (most recent call last):
ogdch_app     |   File "/usr/lib/python2.7/logging/handlers.py", line 76, in emit
ogdch_app     |     if self.shouldRollover(record):
ogdch_app     |   File "/usr/lib/python2.7/logging/handlers.py", line 156, in shouldRollover
ogdch_app     |     msg = "%s\n" % self.format(record)
ogdch_app     |   File "/usr/lib/python2.7/logging/__init__.py", line 741, in format
ogdch_app     |     return fmt.format(record)
ogdch_app     |   File "/usr/lib/python2.7/logging/__init__.py", line 465, in format
ogdch_app     |     record.message = record.getMessage()
ogdch_app     |   File "/usr/lib/python2.7/logging/__init__.py", line 325, in getMessage
ogdch_app     |     msg = str(self.msg)
ogdch_app     |   File "/usr/lib/ckan/venv/src/ckan/ckan/lib/navl/dictization_functions.py", line 19, in __str__
ogdch_app     |     raise Invalid(_('Missing value'))
ogdch_app     | Invalid: Invalid: u'Missing value'
ogdch_app     | Logged from file validators.py, line 158
```

It also happens for all our fields that use that validator: `see_alsos`, `relations`

So what I think we should do is get rid of that validator and replace it by a function that does the work that it is supposed to do and call this function by all the other validators (temporals, contact_point, see_alsos, relations), then build tests for all those validators. 

I think refactoring this would be worth wile, but will take some time and effort.

Reply:
Wait, was this comment meant as a comment to my comment pointing out not throwing an error? I am a little confused 😅

Reply:
But yeah, you are right, I think it is crucial that we get an understanding of the validators first